scm-webapp/src/test/java/sonia/scm/lifecycle/AdminAccountStartupActionTest.java

```diff
@@ -0,0 +1,132 @@
+import sonia.scm.security.PermissionAssigner;
+import sonia.scm.security.PermissionDescriptor;
+import sonia.scm.user.User;
+import sonia.scm.user.UserManager;
+import sonia.scm.user.UserTestData;
+
+import java.util.Collection;
+import java.util.List;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+@ExtendWith(MockitoExtension.class)
+class AdminAccountStartupActionTest {
+
+  @Mock
+  private PasswordService passwordService;
+  @Mock
+  private UserManager userManager;
+  @Mock
+  private PermissionAssigner permissionAssigner;
+
+  @InjectMocks
+  private AdminAccountStartupAction startupAction;
+
+  @Test
+  void shouldCreateAdminAccountIfNoUserExistsAndAssignPermissions() {
+    when(passwordService.encryptPassword("scmadmin")).thenReturn("secret");
+
+    startupAction.run();
+
+    verifyAdminCreated();
+    verifyAdminPermissionsAssigned();
+  }
+
+  @Test
+  void shouldCreateAdminAccountIfOnlyAnonymousUserExistsAndAssignPermissions() {
+    when(userManager.getAll()).thenReturn(Lists.newArrayList(SCMContext.ANONYMOUS));
+    when(userManager.contains(SCMContext.USER_ANONYMOUS)).thenReturn(true);
+    when(passwordService.encryptPassword("scmadmin")).thenReturn("secret");
+
+    startupAction.run();
+
+    verifyAdminCreated();
+    verifyAdminPermissionsAssigned();
+  }
+
+  @Test
+  @MockitoSettings(strictness = Strictness.LENIENT)
+  void shouldSkipAdminAccountCreationIfPropertyIsSet() {
+    System.setProperty("sonia.scm.skipAdminCreation", "true");
+
+    try {
+      startupAction.run();
+
+      verify(userManager, never()).create(any());
+      verify(permissionAssigner, never()).setPermissionsForUser(anyString(), any(Collection.class));
+    } finally {
+      System.setProperty("sonia.scm.skipAdminCreation", "");
+    }
+  }
+
+  @Test
+  void shouldDoNothingOnSecondStart() {
+    List<User> users = Lists.newArrayList(UserTestData.createTrillian());
+    when(userManager.getAll()).thenReturn(users);
+
+    startupAction.run();
+
+    verify(userManager, never()).create(any(User.class));
+    verify(permissionAssigner, never()).setPermissionsForUser(anyString(), any(Collection.class));
+  }
+
+  private void verifyAdminPermissionsAssigned() {
+    ArgumentCaptor<String> usernameCaptor = ArgumentCaptor.forClass(String.class);
+    ArgumentCaptor<Collection<PermissionDescriptor>> permissionCaptor = ArgumentCaptor.forClass(Collection.class);
+    verify(permissionAssigner).setPermissionsForUser(usernameCaptor.capture(), permissionCaptor.capture());
+    String username = usernameCaptor.getValue();
+    assertThat(username).isEqualTo("scmadmin");
+    PermissionDescriptor descriptor = permissionCaptor.getValue().iterator().next();
+    assertThat(descriptor.getValue()).isEqualTo("*");
```

Comment:
Not sure whether this is much better, but... Just an alternative to argument matchers

```suggestion
    verify(permissionAssigner).setPermissionsForUser(
      argThat(username -> { assertThat(username).isEqualTo("scmadmin"); return true; }),
      argThat(descriptor -> { assertThat(descriptor.iterator().next().getValue()).isEqualTo("*"); return true; })
    );
```